File: leetcode/656_hard_Coin_Path.py
class Solution(object):

    def cheapestJump(self, A, B):
        """
        :type A: List[int]
        :type B: int
        :rtype: List[int]
        """
        if not A or A[0] == -1:
            return []
        dp = [[float('inf')] for i in A]
        dp[0] = [A[0], 1]
        N = len(A)
        for i in range(N):
            if A[i] == -1:
                continue
            for j in range(i + 1, min(N, i + B + 1)):
                if A[j] == -1:
                    continue
                if dp[i][0] + A[j] < dp[j][0]:
                    dp[j] = [dp[i][0] + A[j]] + dp[i][1:] + [j + 1]
                if dp[i][0] + A[j] == dp[j][0]:
                    path = dp[i][1:] + [j + 1]
                    if path < dp[j][1:]:
                        dp[j] = [dp[i][0] + A[j]] + dp[i][1:] + [j + 1]
        # return dp
        return dp[-1][1:]

File: leetcode/656_hard_Coin_Path.java
// premium 的题目，题目可以看这里： http://bookshadow.com/weblog/2017/08/06/leetcode-coin-path/

//  modified from yijiezhu 
public class Solution {
    public List<Integer> cheapestJump(int[] A, int B) {
        int n = A.length;
        if (n == 0 || A[n - 1] == -1) 
            return Collections.emptyList();
        
        List<Integer>[] C = (List<Integer>[]) new List[n];  // 存放路径
        int[] D = new int[n];
        Arrays.fill(D, Integer.MAX_VALUE);
        for (int i = 0; i < n; ++i) 
            C[i] = new ArrayList<>();
        
        C[n - 1].add(n);
        D[n - 1] = A[n - 1];
        int end = n - 1;
        
        for (int i = n - 1; i >= 0 && i >= end; --i) {
            if (A[i] == -1) 
                continue;
            for (int j = 1; j <= B && j <= i; ++j) {
                int k = i - j;
                if (A[k] == -1) 
                    continue;
                if (D[i] + A[k] <= D[k]) {
                    C[k] = new ArrayList<>(C[i]);
                    C[k].add(k + 1);
                    end = Math.min(end, k);
                    D[k] = D[i] + A[k];
                    // System.out.println(k + ":" + C[k]);
                }
            }
        }
        if (C[0] == null) 
            C[0] = new ArrayList<>();
        Collections.reverse(C[0]);
        return C[0];
    }
}
